flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManager.java

```diff
@@ -320,5 +333,16 @@ private void internalStopPod(String podName) {
 					}
 				}
 			);
+
+		final KubernetesWorkerNode kubernetesWorkerNode = workerNodes.remove(resourceId);
+		final WorkerResourceSpec workerResourceSpec = podWorkerResources.remove(podName);
+
+		// If the stopped pod is requested in the current attempt (workerResourceSpec is known) and is not yet added,
+		// we need to notify ActiveResourceManager to decrease the pending worker count.
+		if (workerResourceSpec != null && kubernetesWorkerNode == null) {
```

Comment:
I think we don't want to call `notifyNewWorkerAllocationFailed` and decrease the pending count when a recovered pod is being stopped. The `pendingWorkerCounter` only accounts for the workers requested in the current  attempt.

Reply:
But what we want to do is to restart a recovered pod if it fails and it is needed. Assume that our cluster recovers a pod from a previous attempt. Then we submit a job and the job uses slots of this pod. Now the pod fails. Unlike in the case of a started pod, Flink won't try to restart this pod. I think there should not be a difference between in the failure handling between started and recovered pods.

Reply:
I think what we need to restart is the workers that are requested by the slot manager, but not yet registered to the slot manager. Because in such cases, the slot manager does not know that these pods are failed and will keep expecting them to register. I believe that is the behavior before.

For a worker that is already registered, in case of failure:
- If there are slots on the failed worker that are already in use, JM will realize that the slots are lost, and send out new slot requests, which will trigger starting of new workers if needed.
- If none of the slots on the failed worker is in use, then we don't loose anything. New workers will be started when new slot requests are received and cannot be satisfied by registered free slots.

For a pod recovered from the previous attempt, if it is not yet registered, slot manager will not be aware of it at all. Thus, slot manager should already requested enough workers for satisfying the slot requests even without this failed pod. Moreover, we have not increased the `pendingWorkerCounter` for the recovered pod and we cannot do that because we don't know its `workerResourceSpec` until it is registered. So we should also not decrease the `pendingWorkerCounter` when the recovered pod is added/failed, otherwise the `pendingWorkerCounter` will go out of sync.

Reply:
With `requestKubernetesPodIfRequired` now tries to request pods for all the `workerResourceSpec`s, there's not much differences between in failure handling of started/recovered pods. The only difference is whether we decrease the `pendingWorkerCounter` or not.

Reply:
In the current state with calling `requestKubernetesPodIfRequired` it should now work.

I think we are still lacking a bit of test coverage, though. For example if a recovered pod is being used by a job and if the pod now fails, it should be restarted because the `SlotManager` needs the pod.

Reply:
If a recovered pod is being used by a job, there could be two kinds of situations.
- If the slots are already offered to JM, then JM will find out that the TM/slots are lost, and re-request the slots if needed.
- If the slots are already assigned by SM but not yet offered to JM, RM will discover the TM failure and notify JM about the allocation failure. If needed, JM will re-request the slot. This behavior is covered by `SlotManagerImplTest#testNotifyFailedAllocationWhenTaskManagerTerminated`.